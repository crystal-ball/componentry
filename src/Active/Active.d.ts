import * as React from 'react'
import {
  ActiveContainerProps,
  ActiveTriggerProps,
  ActiveContentProps,
} from '../utils/componentry-props'

declare const ActiveContent: React.ComponentType<ActiveContentProps>
declare const ActiveTrigger: React.ComponentType<ActiveTriggerProps>

interface ActiveComponent extends React.ComponentClass<ActiveContainerProps> {
  Content: typeof ActiveContent
  Trigger: typeof ActiveTrigger
}

/** https://crystal-ball.github.io/componentry/components/active */
declare const Active: ActiveComponent

export default Active
