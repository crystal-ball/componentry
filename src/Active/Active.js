// @flow
import withActive from '../withActive'
import activeContainer from '../active-container-factory'
import activeContent from '../active-content-factory'
import activeTrigger from '../active-trigger-factory'

const Content = activeContent({
  arias: { id: true, hidden: true },
  element: 'active',
})

const Trigger = activeTrigger({
  arias: { controls: true },
  element: 'active',
})

const Active = activeContainer({
  Content: withActive(Content),
  Trigger: withActive(Trigger),
  element: 'active',
  escHandler: true,
})

export default Active
