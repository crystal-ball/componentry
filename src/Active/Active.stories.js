/* global module */

import React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import Active from './ActiveComponent'
import Block from '../Block/Block'

storiesOf('Active', module).add('<Active />', () => (
  <Block>
    <h2 className='storybook-section-header'>Uncontrolled component</h2>
    <Active
      onActivate={action('onActivate event')}
      onActivated={action('onActivated event')}
      onDeactivate={action('onDeactivate event')}
      onDeactivated={action('onDeactivated event')}
    >
      <Active.Trigger>Toggle</Active.Trigger>
      <Active.Content>Active component content</Active.Content>
    </Active>

    <h2 className='storybook-section-header'>Controlled</h2>
    <Active activate={action('activate event')}>
      <Active.Trigger>Toggle</Active.Trigger>
      <Active.Content>Active component content</Active.Content>
    </Active>

    <h2 className='storybook-section-header'>Default active override</h2>
    <Active defaultActive Trigger='Toggle' Content='Active component content' />

    <h2 className='storybook-section-header'>Deactivate on external clicks</h2>
    <Active clickEvents Trigger='Toggle' Content='Active component content' />

    <h2 className='storybook-section-header'>Deactivate on esc</h2>
    <Active escEvents Trigger='Toggle' Content='Active component content' />

    <h2 className='storybook-section-header'>Handle mouse events</h2>
    <Active mouseEvents Trigger='Toggle' Content='Active component content' />

    <h2 className='storybook-section-header'>Shorthand subcomponents</h2>
    <Active Trigger='Toggle' Content='Active component content' />

    <h2 className='storybook-section-header'>Size and direction</h2>
    <Active size='sm' direction='up'>
      <Active.Trigger>Toggle</Active.Trigger>
      <Active.Content>Active component content</Active.Content>
    </Active>

    <h2 className='storybook-section-header'>FaCC pattern</h2>
    <Active>
      {({ active, activate, deactivate }) => (
        <>
          <div>{String(active)}</div>
          <button type='button' onClick={activate}>
            activate
          </button>
          <button type='button' onClick={deactivate}>
            deactivate
          </button>
        </>
      )}
    </Active>
  </Block>
))
