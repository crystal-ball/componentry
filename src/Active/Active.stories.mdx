<!-- eslint-disable no-console -->

import { Meta } from '@storybook/addon-docs/blocks'
import { DocsTitle } from '../storybook/DocsTitle'
import { Active } from './Active'

<Meta title='Components/Active' />

<DocsTitle title='Active' experimental />

## Uncontrolled component

<Active
  onActivate={console.log}
  onActivated={console.log}
  onDeactivate={console.log}
  onDeactivated={console.log}
>
  <Active.Action>Toggle</Active.Action>
  <Active.Content>Active component content</Active.Content>
</Active>

## Controlled

<Active activate={console.log}>
  <Active.Action>Toggle</Active.Action>
  <Active.Content>Active component content</Active.Content>
</Active>

## Default active override

<Active Action='Toggle' Content='Active component content' defaultActive />

## Deactivate on external clicks

<Active Action='Toggle' Content='Active component content' clickEvents />

## Deactivate on esc

<Active Action='Toggle' Content='Active component content' escEvents />

## Handle mouse events

<Active Action='Toggle' Content='Active component content' mouseEvents />

## Shorthand subcomponents

<Active Action='Toggle' Content='Active component content' />

## Size and direction

<Active size='sm' direction='up'>
  <Active.Action>Toggle</Active.Action>
  <Active.Content>Active component content</Active.Content>
</Active>

## FaCC pattern

<Active>
  {({ active, activate, deactivate }) => (
    <>
      <div>{String(active)}</div>
      <button type='button' onClick={activate}>
        activate
      </button>
      <button type='button' onClick={deactivate}>
        deactivate
      </button>
    </>
  )}
</Active>
