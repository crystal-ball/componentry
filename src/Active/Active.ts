import { activeContainerBuilder } from '../utils/active-container-component-builder'
import { activeActionBuilder } from '../utils/active-action-component-builder'
import { activeContentBuilder } from '../utils/active-content-component-builder'
import {
  ActiveActionBaseProps,
  ActiveContainerBaseProps,
  ActiveContentBaseProps,
  ComponentBaseProps,
} from '../utils/types'

export interface ActiveProps
  extends ActiveContainerBaseProps,
    ComponentBaseProps<'div'> {}

export interface ActiveActionProps
  extends ActiveActionBaseProps,
    ComponentBaseProps<'button'> {}

export interface ActiveContentProps
  extends ActiveContentBaseProps,
    ComponentBaseProps<'div'> {}

export interface Active {
  (props: ActiveProps): React.ReactElement
  /**
   * [Active action component üìù](https://componentry.design/components/active)
   */
  Action: React.FC<ActiveActionProps>
  /**
   * [Active content component üìù](https://componentry.design/components/active)
   */
  Content: React.FC<ActiveContentProps>
}

/**
 * [Active component üìù](https://componentry.design/components/active)
 * @experimental
 */
export const Active = activeContainerBuilder<ActiveProps>('Active', {
  escEvents: true,
}) as Active

Active.Action = activeActionBuilder<ActiveActionProps>('ActiveAction', {
  aria: { controls: true },
})

Active.Content = activeContentBuilder<ActiveContentProps>('ActiveContent', {
  aria: { id: true, hidden: true },
})
