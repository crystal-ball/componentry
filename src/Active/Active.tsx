import { activeContainerBuilder } from '../utils/active-container-component-builder'
import { activeActionComponent } from '../utils/active-action-component-builder'
import { activeContentComponent } from '../utils/active-content-component-builder'
import { BaseActiveContainerProps, BaseProps } from '../utils/base-types'

export interface ActiveProps
  extends BaseActiveContainerProps,
    BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {}

export interface ActiveActionProps
  extends BaseProps,
    Omit<React.ComponentPropsWithoutRef<'button'>, 'className'> {}

export interface ActiveContentProps
  extends BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {}

export interface Active {
  (props: ActiveProps): React.ReactElement
  /**
   * [Active action component üìù](https://componentry.design/components/active)
   */
  Action: React.FC<ActiveActionProps>
  /**
   * [Active content component üìù](https://componentry.design/components/active)
   */
  Content: React.FC<ActiveContentProps>
}

/**
 * [Active component üìù](https://componentry.design/components/active)
 */
export const Active = activeContainerBuilder<ActiveProps>('Active', {
  escEvents: true,
}) as Active

Active.Action = activeActionComponent<ActiveActionProps>('ActiveAction', {
  aria: { controls: true },
})

Active.Content = activeContentComponent<ActiveContentProps>('ActiveContent', {
  aria: { id: true, hidden: true },
})
