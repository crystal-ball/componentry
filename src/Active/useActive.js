import { useContext, useEffect, useState } from 'react'
import Context from './Context'

/**
 * Components that need transitional active and visible states will use the
 * Consumer to access the activeCtx, and pass Wrapped into the ActiveTransition
 * component, which uses state to manage passing active and visible according
 * to the transition duration for that component.
 */

export const useActive = () => useContext(Context)

/**
 * Hook handles transitioning display and opacity using active and visible
 * state using the following rules:
 *
 * - Show: Set active true to display element, then set visible true to
 *   transition element opacity using css transitions
 * - Hide: Set visible false to start element opacity using css transitions,
 *   then after transition duration set active false to set display none
 */
export const useVisible = (active, duration = 300) => {
  const [state, updateState] = useState({ active, visible: active })

  useEffect(() => {
    if (active) {
      updateState({ active: true, visible: false })
      setTimeout(() => {
        updateState({ active: true, visible: true })
      }, 17)
    } else {
      updateState({ active: true, visible: false })
      setTimeout(() => {
        updateState({ active: false, visible: false })
      }, duration)
    }
  }, [active])

  return state
}
