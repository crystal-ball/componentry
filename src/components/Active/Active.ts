import React from 'react'
import { createActiveAction } from '../../utils/create-active-action-component'
import { createActiveContainer } from '../../utils/create-active-container-component'
import { createActiveContent } from '../../utils/create-active-content-component'
import { UtilityProps } from '../../utils/utility-props'
import { Link } from '../Link/Link'
import {
  ActiveActionBaseProps,
  ActiveContainerBaseProps,
  ActiveContentBaseProps,
} from './active-types'

export interface ActiveProps
  extends ActiveContainerBaseProps,
    UtilityProps,
    React.ComponentPropsWithoutRef<'div'> {}

export interface ActiveActionProps
  extends ActiveActionBaseProps,
    UtilityProps,
    React.ComponentPropsWithoutRef<'button'> {}

export interface ActiveContentProps
  extends ActiveContentBaseProps,
    UtilityProps,
    React.ComponentPropsWithoutRef<'div'> {}

export interface Active {
  (props: ActiveProps): React.ReactElement
  /**
   * [Active action component üìù](https://componentry.design/components/active)
   */
  Action: React.FC<ActiveActionProps>
  /**
   * [Active content component üìù](https://componentry.design/components/active)
   */
  Content: React.FC<ActiveContentProps>
}

/**
 * [Active component üìù](https://componentry.design/components/active)
 * @experimental
 */
export const Active = createActiveContainer('Active', {
  escEvents: true,
}) as Active

Active.Action = createActiveAction('ActiveAction', {
  aria: { controls: true },
  defaultAs: Link,
})

Active.Content = createActiveContent('ActiveContent', {
  aria: { id: true, hidden: true },
})
