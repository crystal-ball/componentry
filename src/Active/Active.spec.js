import React from 'react'
import { render } from '@testing-library/react'

import Active from './ActiveComponent'

import activationTestSuite from '../utils-test/activation-tests'
import elementTests from '../utils-test/element-tests'

describe('<Active />', () => {
  activationTestSuite(Active, { name: 'active' })
  elementTests(Active)
  elementTests(Active.Trigger)
  elementTests(Active.Content)
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Active /> snapshots', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Active>
        <Active.Trigger>Trigger</Active.Trigger>
        <Active.Content>Content</Active.Content>
      </Active>,
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
