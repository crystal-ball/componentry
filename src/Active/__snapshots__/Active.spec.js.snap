// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Active /> snapshots renders correctly 1`] = `
<div
  class="ðŸ…²-active"
  data-id="test-guid"
  data-testid="active"
>
  <button
    aria-controls="test-guid"
    class="ðŸ…²-active-action active-action-primary"
    type="button"
  >
    Action
  </button>
  <div
    aria-hidden="true"
    class="ðŸ…²-active-content active-content-primary"
    id="test-guid"
  >
    Content
  </div>
</div>
`;
